org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandlerTest.java

```diff
@@ -403,8 +430,32 @@ public void testCodeAction_removeUnterminatedString() throws Exception{
 		Assert.assertNotNull(codeActions);
 		Assert.assertFalse(codeActions.isEmpty());
 		Assert.assertEquals(codeActions.get(0).getRight().getKind(), CodeActionKind.QuickFix);
-		Command c = codeActions.get(0).getRight().getCommand();
-		Assert.assertEquals(CodeActionHandler.COMMAND_ID_APPLY_EDIT, c.getCommand());
+		WorkspaceEdit e = (WorkspaceEdit) codeActions.get(0).getRight().getEdit();
+		Assert.assertNotNull(e);
+	}
+
+	@Test
+	public void testCodeActionLiteral_removeUnterminatedString() throws Exception{
+		when(preferenceManager.getClientPreferences().isSupportedCodeActionKind(CodeActionKind.QuickFix)).thenReturn(true);
+		ICompilationUnit unit = getWorkingCopy(
+				"src/java/Foo.java",
+				"public class Foo {\n"+
+						"	void foo() {\n"+
+						"String s = \"some str\n" +
```

Comment:
I've tested kak-lsp on this and it does work. Thanks so much!

There is an interesting issue which I'll try to fix soon in `kak-lsp`:
`jdtls` sends a code action containing both an `edit` and a `command`.
[LSP](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/) says

 > /**
 >  * If a code action provides an edit and a command, first the edit is
 >  * executed and then the command.
 >  */

Which would be wrong here because the `edit` and `command` do the same thing, but they are not idempotent.

However `jdtls` does not include `java.apply.workspaceEdit` in the methods it supports in `workspace/executeCommand` (`executeCommandProvider`).
I guess `kak-lsp` should recognize this and not send `java.apply.workspaceEdit`, thus avoiding the error.
Maybe I'll add a warning to the logs. I don't think LSP specifies what should happen in this case.

Here is the code action sent by `jdtls`:

<details>

```json
{
  "jsonrpc": "2.0",
  "id": 11,
  "result": [
    {
      "title": "Insert missing quote",
      "kind": "quickfix",
      "diagnostics": [
        {
          "range": {
            "start": {
              "line": 5,
              "character": 15
            },
            "end": {
              "line": 5,
              "character": 24
            }
          },
          "severity": 1,
          "code": "1610612995",
          "source": "Java",
          "message": "String literal is not properly closed by a double-quote"
        }
      ],
      "edit": {
        "changes": {},
        "documentChanges": [
          {
            "textDocument": {
              "version": null,
              "uri": "file:///home/johannes/git/kak-lsp/t/java/test.java"
            },
            "edits": [
              {
                "range": {
                  "start": {
                    "line": 5,
                    "character": 24
                  },
                  "end": {
                    "line": 5,
                    "character": 24
                  }
                },
                "newText": "\""
              }
            ]
          }
        ]
      },
      "command": {
        "title": "Insert missing quote",
        "command": "java.apply.workspaceEdit",
        "arguments": [
          {
            "changes": {},
            "documentChanges": [
              {
                "textDocument": {
                  "version": null,
                  "uri": "file:///home/johannes/git/kak-lsp/t/java/test.java"
                },
                "edits": [
                  {
                    "range": {
                      "start": {
                        "line": 5,
                        "character": 24
                      },
                      "end": {
                        "line": 5,
                        "character": 24
                      }
                    },
                    "newText": "\""
                  }
                ]
              }
            ]
          }
        ]
      }
    },
  ]
}
```

</details>

Reply:
There's still a bunch of other places where `jdtls` requires `executeCommand` of `java.apply.workspaceEdit` (for example the "add JavaDoc" code action.
When I add a hack to kak-lsp to support this command (via `workspace/applyEdit`) that means that the code action from the earlier comment will add two quotes instead of just one.

Can we convert those other places to also use `codeAction.edit` instead of only the `java.apply.workspaceEdit` command?

Reply:
I was able to make "Generate toString()" work with this patch but not yet the "Add Javadoc" one.

```diff
diff --git a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/ClientPreferences.java b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/ClientPreferences.java
index 5bfd515a..66852dd0 100644
--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/ClientPreferences.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/ClientPreferences.java
@@ -482,6 +482,12 @@ public class ClientPreferences {
 			&& capabilities.getWorkspace().getWorkspaceEdit().getChangeAnnotationSupport() != null;
     }
 
+	public boolean isWorkspaceEditSupport() {
+		return v3supported
+			&& capabilities.getWorkspace() != null
+			&& capabilities.getWorkspace().getWorkspaceEdit() != null;
+    }
+
 	public boolean skipProjectConfiguration() {
 		return Boolean.parseBoolean(extendedClientCapabilities.getOrDefault("skipProjectConfiguration", "false").toString());
 	}
diff --git a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java
index c4b10686..ddbb4ba9 100644
--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java
@@ -704,8 +704,12 @@ public class SourceAssistProcessor {
 			if (preferenceManager.getClientPreferences().isSupportedCodeActionKind(kind)) {
 				CodeAction codeAction = new CodeAction(name);
 				codeAction.setKind(kind);
-				codeAction.setCommand(command);
-				codeAction.setData(new CodeActionData(null, priority));
+				if (preferenceManager.getClientPreferences().isWorkspaceEditSupport()) {
+					codeAction.setEdit(edit);
+				} else {
+					codeAction.setCommand(command);
+					codeAction.setData(new CodeActionData(null, priority));
+				}
 				codeAction.setDiagnostics(context.getDiagnostics());
 				return Optional.of(Either.forRight(codeAction));
 			} else {
```